File: scripts/lib/blog-series-config.ts
export interface BlogSeriesConfig {
  readonly id: string;
  readonly name: string;
  readonly icon: string;
  readonly author: string;
  readonly baseUrl: string;
  readonly priorityUser: string | undefined;
  readonly navLink: string;
  readonly postTimeUtc: string;
}

const AUTO_BLOG_ZERO: BlogSeriesConfig = {
  id: "auto-blog-zero",
  name: "Auto Blog Zero",
  icon: "🤖",
  author: "[[auto-blog-zero]]",
  baseUrl: "https://bagrounds.org/auto-blog-zero",
  priorityUser: "bagrounds",
  navLink: "[[index|Home]] > [[auto-blog-zero/index|🤖 Auto Blog Zero]]",
  postTimeUtc: "16:00",
};

const CHICKIE_LOO: BlogSeriesConfig = {
  id: "chickie-loo",
  name: "Chickie Loo",
  icon: "🐔",
  author: "[[chickie-loo]]",
  baseUrl: "https://bagrounds.org/chickie-loo",
  priorityUser: "ChickieLoo",
  navLink: "[[index|Home]] > [[chickie-loo/index|🐔 Chickie Loo]]",
  postTimeUtc: "15:00",
};

const THE_PUBLIC_GOOD: BlogSeriesConfig = {
  id: "the-public-good",
  name: "The Public Good",
  icon: "🏛️",
  author: "[[the-public-good]]",
  baseUrl: "https://bagrounds.org/the-public-good",
  priorityUser: "bagrounds",
  navLink: "[[index|Home]] > [[the-public-good/index|🏛️ The Public Good]]",
  postTimeUtc: "17:00",
};

export const BLOG_SERIES: ReadonlyMap<string, BlogSeriesConfig> = new Map([
  [AUTO_BLOG_ZERO.id, AUTO_BLOG_ZERO],
  [CHICKIE_LOO.id, CHICKIE_LOO],
  [THE_PUBLIC_GOOD.id, THE_PUBLIC_GOOD],
]);

export const lookupSeries = (seriesId: string): BlogSeriesConfig => {
  const series = BLOG_SERIES.get(seriesId);
  if (!series) throw new Error(`Unknown blog series: ${seriesId}. Available: ${[...BLOG_SERIES.keys()].join(", ")}`);
  return series;
};

File: haskell/src/Automation/BlogSeriesConfig.hs
module Automation.BlogSeriesConfig
  ( BlogSeriesConfig (..)
  , lookupSeriesIn
  , extraContentDirectories
  , backfillContentIdsFrom
  , imageBackfillContentDirectoriesFrom
  ) where

import Data.Map.Strict (Map)
import qualified Data.Map.Strict as Map
import Data.Text (Text)
import Data.Time.LocalTime (TimeOfDay)

import Automation.BlogImage.ContentDirectory (ContentDirectory (..), contentDirectoryFromText, contentDirectoryToText)
import Automation.ContextQuery (ContextQuery)

data BlogSeriesConfig = BlogSeriesConfig
  { identifier      :: Text
  , name            :: Text
  , icon            :: Text
  , author          :: Text
  , baseUrl         :: Text
  , priorityUser    :: Maybe Text
  , navLink         :: Text
  , scheduleTime    :: TimeOfDay
  , contextQueries  :: [ContextQuery]
  } deriving (Show, Eq)

lookupSeriesIn :: Map Text BlogSeriesConfig -> Text -> Either Text BlogSeriesConfig
lookupSeriesIn seriesMap seriesId =
  maybe (Left errorMessage) Right (Map.lookup seriesId seriesMap)
  where
    errorMessage = "Unknown blog series: " <> seriesId
      <> ". Available: " <> mconcat (fmap (<> " ") (Map.keys seriesMap))

extraContentDirectories :: [ContentDirectory]
extraContentDirectories = [Reflections, AiBlog]

backfillContentIdsFrom :: [BlogSeriesConfig] -> [Text]
backfillContentIdsFrom allSeries =
  fmap contentDirectoryToText extraContentDirectories <> fmap identifier allSeries

libraryContentDirectories :: [ContentDirectory]
libraryContentDirectories =
  [ Articles, Books, BotChats, Games
  , Products, Software, Tools, Topics
  ]

imageBackfillContentDirectoriesFrom :: [BlogSeriesConfig] -> [ContentDirectory]
imageBackfillContentDirectoriesFrom allSeries =
  extraContentDirectories
    <> fmap (contentDirectoryFromText . identifier) allSeries
    <> libraryContentDirectories
